docs/BLOCK_DEVICE_LOCKING.md

```diff
@@ -0,0 +1,60 @@
+# Locking Block Device Access
+
+*TL;DR: Use BSD file locks
+[(`flock(2)`)](http://man7.org/linux/man-pages/man2/flock.2.html) on block
+device nodes to synchronize access for partitioning and file system formatting
+tools.*
+
+`systemd-udevd` probes all block devices showing up for file system superblock
+and partition table information (utilizing `libblkid`). If another program
+concurrently modifies a superblock or partition table this probing might be
+affected, which is bad in itself, but also might in turn result in undesired
+effects in programs subscribing to `udev` events.
+
+Applications manipulating a block device can temporarily stop `systemd-udevd`
+from processing rules on it — and thus bar it from probing the device — by
+taking a BSD file lock on the block device node. Specifically, whenever
+`systemd-udevd` starts processing a block device it takes a `LOCK_SH|LOCK_NB`
+lock using [(`flock(2)`)](http://man7.org/linux/man-pages/man2/flock.2.html) on
+the main block device (i.e. never on any partition block device, but on the
+device the partition belongs to). If this lock cannot be taken (i.e. `flock()`
+returns `EBUSY`), it refrains from processing the device. If it manages to take
+the lock it is kept for the entire time the device is processed.
+
+Note that `systemd-udevd` also watches all block device nodes it manages for
+`inotify()` `IN_CLOSE` events: whenever such an event is seen, this is used as
+trigger to re-run the rule-set for the device.
+
+These two concepts allow tools such as disk partitioners or file system
+formatting tools to safely and easily take exclusive ownership of a block
+device while operating: before starting work on the block device, they should
+take an `LOCK_EX` lock on it. This has two effects: first of all, in case
+`systemd-udevd` still processes the device the tool will wait for it to
+finish. Secondly, after the lock is taken, it can be sure that that
+`systemd-udevd` will refrain from processing the block device again, and thus
+all other client applications subscribed to it won't get device notifications
+from potentially half-written data either. After the operation is complete the
+partitioner/formatter can simply close the device node again. This has two
+effects: it implicitly releases the lock, so that `systemd-udevd` can process
+events on the device node again. Secondly, it results in an `IN_CLOSE` event,
+which causes `systemd-udevd` to immediately re-process the device, seeing all
+changes the tool made, and notifying subscribed clients about it.
+
+Besides synchronizing block device access between `systemd-udevd` and such
+tools this scheme may also be used to synchronize access between those tools
+among each other, as the established Linux file system locking protocol is
```

Comment:
"among each other" → "themselves"

", as the established Linux file system locking protocol is simply reused for managing block device node access" → "" (I think it's more relevant that this is the same protocol that is used by udev, and the fact that this is a popular Linux protocol is not terribly important)